LR_Modules/newdq.f90

```diff
@@ -78,53 +80,72 @@ subroutine newdq (dvscf, npe)
-                                eigts2(mill(2,ig),na) * &
-                                eigts3(mill(3,ig),na) * &
-                                eigqts(na)
-                          do is = 1, nspin_mag
-                             z1(is) = z1(is) + conjg(z2) * aux2(ig, is)
-                          enddo
-                       enddo !ig
-!$omp end parallel do
-                       do is = 1, nspin_mag
-                          int3(ih,jh,na,is,ipert) = omega * z1(is)
-                       enddo
-                    endif
-                 enddo
-              enddo
-           enddo
-           do na = 1, nat
-              if (ityp(na) == nt) then
-                 !
-                 !    We use the symmetry properties of the ps factor
-                 !
-                 do ih = 1, nh (nt)
-                    do jh = ih, nh (nt)
-                       do is = 1, nspin_mag
-                          int3(jh,ih,na,is,ipert) = int3(ih,jh,na,is,ipert)
-                       enddo
-                    enddo
-                 enddo
-              endif
-           enddo
-        endif
-     enddo
+           !======================================================================= my changes below
+           ijh = 0
+           ! composite index for ih and jh (ksi and ksi')
+           nij = nh(nt)*(nh(nt)+1)/2 ! max number of (ih,jh) pairs per atom type nt
+           N_nt = 0 ! number of atoms of type nt
+           DO na = 1, nat
+              IF ( ityp(na) == nt ) N_nt = N_nt + 1
+           ENDDO
+
+           allocate (tmp(ngm, N_nt))
+           allocate (res1(N_nt, nij))
+           allocate (res2(nij, N_nt))
+           res1(:,:) = (0.d0, 0.d0)
+           res2(:,:) = (0.d0, 0.d0)
+
+           do is = 1, nspin_mag ! loop over spins
+               na_ = 0 ! count atoms of type nt
+               !=============== compute potential (aux2) * phase factors
+               do na = 1, nat ! loop over all atoms
+                  if (ityp(na) == nt) then
+                     na_ = na_ + 1
+                     do ig = 1, ngm ! loop over G-vectors
+                         tmp(ig, na_) = aux2(ig, is) * CONJG( eigts1(mill(1,ig),na) * &
+                                                              eigts2(mill(2,ig),na) * &
+                                                              eigts3(mill(3,ig),na) * &
+                                                              eigqts(na) )
+                     enddo
+                  endif
+               enddo
+               !=============== compute Q*V for all atoms of type nt
+               !    DGEMM( TA,  TB,    M,   N,   K, ALPHA,   A, LDA,                        B, LDB,  BETA,   C,  LDC)
+!              call ZGEMM('T', 'N', N_nt, nij, ngm, 1.0d0, tmp, ngm, CONJG(atom_type(nt)%qpw), ngm, 0.0d0, res1, N_nt)
+               call ZGEMM('C', 'N', nij, N_nt, ngm, 1.0d0, atom_type(nt)%qpw, ngm, tmp, ngm, 0.0d0, res2, nij)
```

Comment:
for zgemm you probaly need to pass dcmplx(0.d0, 0.d0) - complex alpha and beta numbers